src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/Gen2GcCallback.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// NOTE: This code is derived from an implementation originally in dotnet/runtime:
+// https://github.com/dotnet/runtime/blob/v5.0.4/src/libraries/System.Private.CoreLib/src/System/Gen2GcCallback.cs
+//
+// See the commentary in https://github.com/dotnet/roslyn/pull/50156 for notes on incorporating changes made to the
+// reference implementation.
+
+using System.Diagnostics;
+using System.Runtime.ConstrainedExecution;
+using System.Runtime.InteropServices;
+
+namespace System
+{
+    /// <summary>
+    /// Schedules a callback roughly every gen 2 GC (you may see a Gen 0 an Gen 1 but only once)
+    /// (We can fix this by capturing the Gen 2 count at startup and testing, but I mostly don't care)
+    /// </summary>
+    internal sealed class Gen2GcCallback : CriticalFinalizerObject
+    {
+        private readonly Func<bool>? _callback0;
+        private readonly Func<object, bool>? _callback1;
```

Comment:
tbh i would doc these as to how they're different.

Reply:
oh.  i got it now.  what was super confusing was the user of 0 and 1.  which i though had to do with Gen0 and Gen1!  might be good to name differently :)

Reply:
➡️ Keeping this one as-is to minimize deviations from dotnet/runtime